### PR TITLE
Feat: Add support for multiple gateway traits

### PR DESCRIPTION
**Pull request by Kolossi**
### Summary
📝🛠️🚀

This pull request improves the `gateway` trait definition for both `vela-core` and `vela-minimal` charts by adding new parameters and logic to support multiple traits per component and to customize the ingress name and path type. It also removes some empty lines from the README files and updates the status and health policy fields for the trait.

> _This pull request has a goal_
> _To enhance the `gateway` trait's role_
> _It adds some parameters_
> _To customize the ingress factors_
> _And removes empty lines from the whole_

### Walkthrough
*  Add logic and parameters to support multiple gateway traits on a single component and to allow users to specify different path types for the gateway trait ([link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-ea69c9becd88df50fc43ae282de36f9655446ef2fa834330853f154d31737e1cL17-R25),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-ea69c9becd88df50fc43ae282de36f9655446ef2fa834330853f154d31737e1cL44-R34),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-ea69c9becd88df50fc43ae282de36f9655446ef2fa834330853f154d31737e1cL73-R63),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-ea69c9becd88df50fc43ae282de36f9655446ef2fa834330853f154d31737e1cL110-R146),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-3b25468eb9839ab3e8389ab65162933ab669d93249578b75707f7c64ecedde89L17-R25),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-3b25468eb9839ab3e8389ab65162933ab669d93249578b75707f7c64ecedde89L44-R34),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-3b25468eb9839ab3e8389ab65162933ab669d93249578b75707f7c64ecedde89L73-R63),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-3b25468eb9839ab3e8389ab65162933ab669d93249578b75707f7c64ecedde89L110-R146),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0L12-R50),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0L43-R63),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0L69-R72),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0L98-R101),[link](https://github.com/kubevela/kubevela/pull/5860/files?diff=unified&w=0#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0R138-R143)). These changes affect the `schematic.cue.template` field in `charts/vela-core/templates/defwithtemplate/gateway.yaml` and `charts/vela-minimal/templates/defwithtemplate/gateway.yaml`, and the `status.customStatus`, `status.healthPolicy`, `template`, and `parameter` fields in `vela-templates/definitions/internal/trait/gateway.cue`.

By allowing the user to specify a `name` property, supports multiple gateways on a component.  Example use case - service has different internal and external hostnames.

If the gateway `name` is not specified, the behaviour is not changed from the current implementation.

If a name is specified, the updated `customStatus` supports correct status reporting of the relevant ingress in each gateway trait status.

This PR also adds the ability to select the ingress `pathType` via another new property, between "ImplementationSpecfic" (the current hardcoded type, and the new default), "Prefix" and "Exact".

I have discussed this change [in the slack channel](https://cloud-native.slack.com/archives/C01BLQ3HTJA/p1679991847280929) (link corrected from that in initial PR submission) and ~~@wonderflow~~ Da Yin is supportive of me raising this PR.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

I would have liked to have added unit tests, but that is beyond my capability in the kubevela repo.

In terms of my testing, I've used this same code as a custom gateway with this code as both a single and dual gateway trait on a webservice and it works without issue for our organisation.

### Special notes for your reviewer

- The only file I changed was vela-templates/definitions/internal/trait/gateway.cue, the others were changed by the `make && make reviewable` run, so I committed them too.

- I removed the [outputs:service section](https://github.com/kubevela/kubevela/compare/master...Kolossi:kubevela:multi_gateway#diff-eb6e43f0e2cf800a96e1dac66f3f4871067f865d5c694d9efa5c84cb2529a5a0L43-L56) as I believe from my investigation this is immedately dropped due to the following `outputs:ingress` assignment and the service used is in fact created in the webservice component.  Please alert me if this is incorrect.

- I looked in kubevela.io repo for where the gateway docs are so I could edit and raise a PR there, but couldn't see them - are they just auto-generated from the comments in the cue file?

### CODE REVIEW

File: charts/vela-minimal/templates/defwithtemplate/gateway.yaml
Comment (FogDong) [inline charts/vela-minimal/templates/defwithtemplate/gateway.yaml:121]: You can use double if to simplify like:

```
if igs != _|_ if igs.ip != _|_ {
}
```

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [inline vela-templates/definitions/internal/trait/gateway.cue:63]: If you don't need to add an extra string to a CUE var, you can simplify this with brackets only like:

```
outputs: (ingressOutputName)
```

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [inline vela-templates/definitions/internal/trait/gateway.cue:50]: The valid field in `healthPolicy` is `isHealth`, so we can simplify this to:

```
isHealth: ig.status.loadBalancer.ingress[0].ip != _|_
```

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [inline vela-templates/definitions/internal/trait/gateway.cue:18]: There may be some `_|_ bottom error` if ig.status.loadBalancer is `_|_`.

**Review by Somefive — COMMENTED** [review]
Generally LGTM, thanks for the PR! I found that the unit-test reports errors when testing gateway trait with https protocol and there might be some format issues related to the `check-diff` CI. You can run `make reviewable` to fix that.

**Comment by Kolossi** [comment]
> Generally LGTM, thanks for the PR! I found that the unit-test reports errors when testing gateway trait with https protocol and there might be some format issues related to the `check-diff` CI. You can run `make reviewable` to fix that.

I did run "make reviewable" and included the fixes in the PR, so I'm unsure what else to resolve that?

**Comment by Kolossi** [comment]
Thanks @FogDong for that valued feedback.

I'm not going to able to make those suggested changes for  a week or so.  So can we leave the PR open and I'll do the fixes as soon as able.

**Comment by Somefive** [comment]
> > Generally LGTM, thanks for the PR! I found that the unit-test reports errors when testing gateway trait with https protocol and there might be some format issues related to the `check-diff` CI. You can run `make reviewable` to do to fix that.
> 
> I did run "make reviewable" and included the fixes in the PR, so I'm unsure what else to resolve that?

I see. There is an implicit upgrade for the version of readme-generator in the CI process. So you can try to follow the instruction https://github.com/bitnami-labs/readme-generator-for-helm and upgrade your readme-generator.

**Comment by Somefive** [comment]
@Kolossi BTW, you can try to rebase the master branch and it will reduce some outdated tests and make your CI process easier.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue]: changed in latest commit

**Comment by Kolossi** [comment]
> @Kolossi BTW, you can try to rebase the master branch and it will reduce some outdated tests and make your CI process easier.

I've merged latest master and added an explicit install of latest readme-generator-for-helm to [my builder dockerfile](https://www.kolossi.co.uk/kubevela-pr-build/) then rerun `make && make reviewable` though after the merge from master, it didn't seem to change much.

I've done the one fix I can from @FogDong comments, awaiting help with the other 2, so not ready to merge yet.  But has this improved the things you saw @Somefive ?

**Comment by Somefive** [comment]
@Kl
> I've merged latest master and added an explicit install of latest readme-generator-for-helm to [my builder dockerfile](https://www.kolossi.co.uk/kubevela-pr-build/) then rerun make && make reviewable though after the merge from master, it didn't seem to change much.

You see the "Go / check-diff" CI now pass. I think it now works.

> But has this improved the things you saw @Somefive ?

I'm okay with this PR now. Just seems we might need another "rebase". Let me see if I can make some post fix based on your PR.

**Comment by Somefive** [comment]
BTW, usually we prefer to `git rebase master` instead of `git merge master`. Let me see if I can fix it in your case.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [reply vela-templates/definitions/internal/trait/gateway.cue:18]: Currently, CUE can not support short-circuit in if comprehension, refer to issue: https://github.com/cue-lang/cue/issues/2232

Instead of using let, we can combine defaut value and if comprehension like:

```
igs: *null | string
if ig != _|_ if ig.status != _|_ if ig.status.loadbalancer != _|_ {
  igs: ig.status.loadbalancer.ingress[0]
}
```

File: charts/vela-minimal/templates/defwithtemplate/gateway.yaml
Comment (FogDong) [reply charts/vela-minimal/templates/defwithtemplate/gateway.yaml]: Yeah I think it's good with individual lines of ifs as long as we're not using `&&` for short-circuit.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue:18]: ok.  can you explain the difference in cue between `let var=` and just rendering the item as `var:` ?

**Comment by Kolossi** [comment]
> BTW, usually we prefer to `git rebase master` instead of `git merge master`. Let me see if I can fix it in your case.
> 
> Updated: with `git rebase master` (there might be some minor manual fixes during the rebase), you will have limited changes to the master branch as shown [Somefive#3](https://github.com/Somefive/kubevela/pull/3). You can rebase the master branch by yourself or continue working on https://github.com/Somefive/kubevela/tree/multi_gateway_fixed this branch (to replace the branch here).

OK, I prefer rebase too, but I wasn't sure how it might affect the code comments already made.  I'll sort this out as a rebase on to latest master and then add the changes suggested.

**Comment by Kolossi** [comment]
@Somefive @FogDong

FYI I need to rerun `make reviewable` after my local rebase and update to fix PR comments, which I can't do until tomorrow, as it never completes on my work laptop, I have to run it on my home PC.  Once this is done I'll update the PR.

But in the PR description I had a special note for reviewer:

> I looked in kubevela.io repo for where the gateway docs are so I could edit and raise a PR there, but couldn't see them - are they just auto-generated from the comments in the cue file?

I'd like to make sure that the docs are there to explain this usage, but as mentioned I don't seem to be able to find them.  Please advise?

**Comment by Kolossi** [comment]
@Somefive @FogDong 

The automation is reporting [unit tests failing](https://github.com/kubevela/kubevela/actions/runs/4788921912/jobs/8516166441?pr=5860#step:8:9755) at https://github.com/kubevela/kubevela/blob/master/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go#L3850 but it's beyond me to find out what in my changes is causing that.  Would you advise please?

**Comment by Somefive** [comment]
> @Somefive @FogDong
> 
> FYI I need to rerun `make reviewable` after my local rebase and update to fix PR comments, which I can't do until tomorrow, as it never completes on my work laptop, I have to run it on my home PC. Once this is done I'll update the PR.
> 
> But in the PR description I had a special note for reviewer:
> 
> > I looked in kubevela.io repo for where the gateway docs are so I could edit and raise a PR there, but couldn't see them - are they just auto-generated from the comments in the cue file?
> 
> I'd like to make sure that the docs are there to explain this usage, but as mentioned I don't seem to be able to find them. Please advise?

Yes, the doc there is auto-generated, using the CLI. You can check out this doc https://kubevela.io/docs/contributor/cli-ref-doc.

**Comment by Somefive** [comment]
> @Somefive @FogDong
> 
> The automation is reporting [unit tests failing](https://github.com/kubevela/kubevela/actions/runs/4788921912/jobs/8516166441?pr=5860#step:8:9755) at https://github.com/kubevela/kubevela/blob/master/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go#L3850 but it's beyond me to find out what in my changes is causing that. Would you advise please?

Some of the tests are flaky tests which might be affected by the environment, like the network or some rate limitation. You can ping me and I can help you to retest that. FYI, we are planning to make some auto bots so that contributors can trigger rerun by commenting like `/retest`.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [reply vela-templates/definitions/internal/trait/gateway.cue:18]: We can set default value, value or type with `:`, but only value can be set in `let`.

**Comment by Kolossi** [comment]
@FogDong @Somefive 

I've fixed up the PR to be rebased rather than merged to latest changes.

The unittest failed, but somewhere else now, so perhaps just down to the "flaky" tests?

So I believe I've now addressed the outstanding issues, this should be good to merge if you approve?

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue]: So I wonder now why to ever use `let var=`?

I guess just that it doesn't appear as an output?

**Comment by Somefive** [comment]
@Kolossi The rebase looks fairly great! BTW, we have just removed the `vela-minimal` in charts/ on the master branch. You can remove that folder in your commit. For the UnitTest, it is a bit weird (it seems unrelated to your commit but also not a known flaky test for me). I will checkout your branch and try to run that test locally and see if there is anything broken. I will also rerun the test as well.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (FogDong) [reply vela-templates/definitions/internal/trait/gateway.cue]: Yes `let` is a declaration so the let won't be evaluated in the final result of CUE.
In your cases, we'll only lookup the value in `isHealth` and `customStatus`, so there's no differences whether you're using `let` or not.

Here's some official doc in CUE explaining let: https://cuelang.org/docs/references/spec/#let-declarations

**Comment by Kolossi** [comment]
> BTW, we have just removed the `vela-minimal` in charts/ on the master branch. You can remove that folder in your commit.

Done!

**Comment by Kolossi** [comment]
> @Kolossi I saw one of the unit test failed and it seems to be related to the gateway trait modification. Could you help check if it is caused by the change of this PR?
> 
> > [Fail] Test Application Controller [It] test application with trait-gateway support https protocol
> > /home/runner/work/kubevela/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:3850

Sorry, it's beyond my level of knowledge to look into that.  :-(

**Comment by Somefive** [comment]
> > @Kolossi I saw one of the unit test failed and it seems to be related to the gateway trait modification. Could you help check if it is caused by the change of this PR?
> > 
> > > [Fail] Test Application Controller [It] test application with trait-gateway support https protocol
> > > /home/runner/work/kubevela/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:3850
> 
> Sorry, it's beyond my level of knowledge to look into that.  :-(

Okay, I will try to checkout to this commit and see what happens to that unit test. It might take some time.

File: vela-templates/definitions/internal/trait/gateway.cue
Comment (chivalryq) [inline vela-templates/definitions/internal/trait/gateway.cue:56]: Hi @Kolossi I just checked out the branch and checked the unit test. 

In this test case: "test application with trait-gateway support https protocol". It will test a app with gateway trait. and since you change the ingress healthPolicy to check `.status.loadBalancer.ingress` field of Ingress which is not generated in unit test environment. 

I add some code to check the process:

```go

		bs, _ := yaml.Marshal(curApp)
		fmt.Fprintf(GinkgoWriter, "curApp: \n%s\n", string(bs))

		ing := &networkingv1.Ingress{}
		err := k8sClient.Get(ctx, client.ObjectKey{Name: "myworker", Namespace: "vela-test-with-trait-gateway-https"}, ing)
		Expect(err).Should(BeNil())
		bs, _ = yaml.Marshal(ing)
		fmt.Fprintf(GinkgoWriter, "ingress: \n%s\n", string(bs))

```

You can run unit test locally to see details.
```shell
ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application
 ```

My suggestion here is to add a gateway trait in `testdata` and use that. Keep the healthPolicy simple and make it pass the workflow when there's no ingress controller.

Comment (chivalryq) [reply vela-templates/definitions/internal/trait/gateway.cue:56]: @Kolossi Yes. I think it's OK to leave the built-in gateway trait's behavior like that as the same before (Better because it'll check the Ingress.) 

Now I got the second requirements to link multiple domains with different certs to one service.

Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue:56]: OK, great I will do the unittest checks today.

> Now I got the second requirements to link multiple domains with different certs to one service.

This is the requirement that had me implement this, but I didn't quite get this comment.  Are you saying this needs work to achieve your second requirement, or you want this merged soon as it now meets a requirement you have? :-)

Comment (chivalryq) [reply vela-templates/definitions/internal/trait/gateway.cue:56]: The latter. :D It's a common case I think.

Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue:56]: I tried doing `make && make test` in my docker-based builder, but this failed in `./pkg/definition/gen_sdk/gen_skd_test.go` since `exec: "docker": executable file not found in $PATH`.  It's running _in_ docker, it doesn't _have_ docker :-)

Anyway I tried your suggested command `ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application` but this fails apparently with a ginko syntax error as the helptext is displayed:

```
root@cb45868a70c6:~/kubevela# ginkgo -focus "test application with trait-gateway support https protocol" ./pkg/controller/core.oam.dev/v1alpha2/application
flag provided but not defined: -ginkgo.timeout
Usage of /root/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/application.test:
  -ginkgo.debug
        If set, ginkgo will emit node output to files when running in parallel.
  -ginkgo.dryRun
        If set, ginkgo will walk the test hierarchy without actually running anything.  Best paired with -v.
  -ginkgo.failFast
        If set, ginkgo will stop running a test suite after a failure occurs.
  -ginkgo.failOnPending
        If set, ginkgo will mark the test suite as failed if any specs are pending.
  -ginkgo.flakeAttempts int
        Make up to this many attempts to run each spec. Please note that if any of the attempts succeed, the suite will not be failed. But any failures will still be recorded. (default 1)
  -ginkgo.focus value
        If set, ginkgo will only run specs that match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.noColor
        If set, suppress color output in default reporter.
  -ginkgo.noisyPendings
        If set, default reporter will shout about pending tests. (default true)
  -ginkgo.noisySkippings
        If set, default reporter will shout about skipping tests. (default true)
  -ginkgo.parallel.node int
        This worker node's (one-indexed) node number.  For running specs in parallel. (default 1)
  -ginkgo.parallel.streamhost string
        The address for the server that the running nodes should stream data to.
  -ginkgo.parallel.synchost string
        The address for the server that will synchronize the running nodes.
  -ginkgo.parallel.total int
        The total number of worker nodes.  For running specs in parallel. (default 1)
  -ginkgo.progress
        If set, ginkgo will emit progress information as each spec runs to the GinkgoWriter.
  -ginkgo.randomizeAllSpecs
        If set, ginkgo will randomize all specs together.  By default, ginkgo only randomizes the top level Describe, Context and When groups.
  -ginkgo.regexScansFilePath
        If set, ginkgo regex matching also will look at the file path (code location).
  -ginkgo.reportFile string
        Override the default reporter output file path.
  -ginkgo.reportPassed
        If set, default reporter prints out captured output of passed tests.
  -ginkgo.seed int
        The seed used to randomize the spec suite. (default 1683539470)
  -ginkgo.skip value
        If set, ginkgo will only run specs that do not match this regular expression. Can be specified multiple times, values are ORed.
  -ginkgo.skipMeasurements
        If set, ginkgo will skip any measurement specs.
  -ginkgo.slowSpecThreshold float
        (in seconds) Specs that take longer to run than this threshold are flagged as slow by the default reporter. (default 5)
  -ginkgo.succinct
        If set, default reporter prints out a very succinct report
  -ginkgo.trace
        If set, default reporter prints out the full stack trace when a failure occurs
  -ginkgo.v
        If set, default reporter print out all specs as they begin.
  -kubeconfig string
        Paths to a kubeconfig. Only required if out-of-cluster.
  -test.bench regexp
        run only benchmarks matching regexp
  -test.benchmem
        print memory allocations for benchmarks
  -test.benchtime d
        run each benchmark for duration d (default 1s)
  -test.blockprofile file
        write a goroutine blocking profile to file
  -test.blockprofilerate rate
        set blocking profile rate (see runtime.SetBlockProfileRate) (default 1)
  -test.count n
        run tests and benchmarks n times (default 1)
  -test.coverprofile file
        write a coverage profile to file
  -test.cpu list
        comma-separated list of cpu counts to run each test with
  -test.cpuprofile file
        write a cpu profile to file
  -test.failfast
        do not start new tests after the first test failure
  -test.fuzz regexp
        run the fuzz test matching regexp
  -test.fuzzcachedir string
        directory where interesting fuzzing inputs are stored (for use only by cmd/go)
  -test.fuzzminimizetime value
        time to spend minimizing a value after finding a failing input (default 1m0s)
  -test.fuzztime value
        time to spend fuzzing; default is to run indefinitely
  -test.fuzzworker
        coordinate with the parent process to fuzz random values (for use only by cmd/go)
  -test.list regexp
        list tests, examples, and benchmarks matching regexp then exit
  -test.memprofile file
        write an allocation profile to file
  -test.memprofilerate rate
        set memory allocation profiling rate (see runtime.MemProfileRate)
  -test.mutexprofile string
        write a mutex contention profile to the named file after execution
  -test.mutexprofilefraction int
        if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
  -test.outputdir dir
        write profiles to dir
  -test.paniconexit0
        panic on call to os.Exit(0)
  -test.parallel n
        run at most n tests in parallel (default 8)
  -test.run regexp
        run only tests and examples matching regexp
  -test.short
        run smaller test suite to save time
  -test.shuffle string
        randomize the execution order of tests and benchmarks (default "off")
  -test.testlogfile file
        write test action log to file (for use only by cmd/go)
  -test.timeout d
        panic test binary after duration d (default 0, timeout disabled)
  -test.trace file
        write an execution trace to file
  -test.v
        verbose: print additional output

Ginkgo ran 1 suite in 8.1917425s

Test Suite Failed
```

I just don't have a dev setup ready to test all this I'm afraid.

I'll update the health as suggested in the previous comment and repush.

Comment (chivalryq) [reply vela-templates/definitions/internal/trait/gateway.cue]: 1. There's a SDK test depends on Docker which is hard to remove. You can simply run a single test. And it takes so long to run the whole test.
2. Check the ginkgo version maybe? I'm using `Ginkgo Version 1.16.5`
3. Another question, have you tested this definition in the component with multiple gateway trait?

Comment (Kolossi) [reply vela-templates/definitions/internal/trait/gateway.cue]: > 2. Check the ginkgo version maybe? I'm using Ginkgo Version 1.16.5

The [instructions by wonderflow](https://github.com/wonderflow/kubevela.io/blob/b4b7bae0a90e0b087df79e5ba5c46fdca072e4f6/docs/contributor/code-contribute.md#run-kubevela-locally) have it installing v2 latest.  Mine picked up v2.9.2!!!

> 3. Another question, have you tested this definition in the component with multiple gateway trait?

Well we corrected a few things in this PR, so actually I haven't used this particular version. It does pass the tests for single usage so I'd expect it to be ok, but I'll test this at some point tomorrow. I'll post the dry-run result to demonstrate.

**Review by Somefive — APPROVED** [review]
Looks great! Thanks a lot! /cc @FogDong

**Comment by chivalryq** [comment]
There seems to be a break change that won't creating Service? @Kolossi

**Comment by chivalryq** [comment]
@Kolossi In your case, the Service is created because `port` in component is filled. Before it can be created by `gateway` trait automatically. So there could be a situation someone is using component without `port` and `gateway` trait. After upgrade it'll fail due to lack of Service.

**Review by FogDong — APPROVED** [review]

**Comment by FogDong** [comment]
Does this PR need to be backported?

**Comment by chivalryq** [comment]
> Does this PR need to be backported?
Let's keep it in v1.9. I think it's a break change for gateway users. Although I think it's ok to remove output:service part to clarify: 
- This trait should be used when need "External Access". Under the hood it creates only Ingress.
- `gateway` can be used together with `expose` (This is mentioned in Chinese community)

**Review by chivalryq — APPROVED** [review]

File: charts/vela-core/templates/defwithtemplate/gateway.yaml
Comment (chivalryq) [inline charts/vela-core/templates/defwithtemplate/gateway.yaml:17]: Wait a minute, what's this "2" It seems a problem caused by generation process

Comment (chivalryq) [reply charts/vela-core/templates/defwithtemplate/gateway.yaml:17]: My bad. It's a YAML grammar. I don't know that.

**Comment by chivalryq** [comment]
@Kolossi Check this app.

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: vela-app
spec:
  components:
    - name: vela-app
      type: webservice
      properties:
        image: oamdev/hello-world
      traits:
        - type: gateway
          properties:
            name: internal-no-svc
            domain: my.internal.name.com
            http:
              "/": 80

```

**Comment by Kolossi** [comment]
Hmm ok, it is as you say a breaking change which isn't what I wanted to happen.  This PR has been merged, but should I make a new PR with a fix for this breaking change? @chivalryq @FogDong ?

**Comment by chivalryq** [comment]
Yes, I am not intended to make a break change. Roll it back +1.

**Comment by Kolossi** [comment]
OK. I've prepared an updated gateway.yml that puts the service capability back in.

I had a misunderstanding with the two `outputs: {...}` entries.

It seemed to me it would be overwritten, I was thinking it was like:

```
outputs: [1]
outputs: [2]
```

and so "obviously" the result would be
```
outputs: [2]
```

But of course cue is different, so the result was
```
outputs: [1,2]
```

I have it in my fork, but are we going to roll back this PR, or should I just make another with the fix and we can move forward without the rollback.

If the latter, I will need to wait until tomorrow to do the `make reviewable` to be sure it's ok.

**Comment by chivalryq** [comment]
Sending a new PR is fine.

**Comment by Kolossi** [comment]
https://github.com/kubevela/kubevela/pull/5978 created

**Pull request merged**
